### ISSUE TITLE
Missing required property writer.staging.dir

### ISSUE THREAD

**Issue by Evangileon** [linkedin/gobblin#292]
Gobblin MapReduce job- Pull from Kafka to HDFS failed while in initialization.
I have 1 maser and 6 slave nodes for HDFS and YARN.
I am sure writer.staging.dir is present in both Gobblin system configuration file and job configuration file. And they are consistent.
Here is the error message:
```
WARN [KafkaSource] Previous offset for partition tianjin-smart-city-1:0 does not exist. This partition will start from the earliest offset: 68333550
WARN [KafkaSource] Previous offset for partition tianjin-smart-city-1:1 does not exist. This partition will start from the earliest offset: 0
WARN [KafkaSource] Previous offset for partition tianjin-smart-city-1:2 does not exist. This partition will start from the earliest offset: 0
WARN [KafkaSource] Previous offset for partition tianjin-smart-city-1:3 does not exist. This partition will start from the earliest offset: 0
WARN [KafkaSource] Avg event size for partition tianjin-smart-city-1:0 not available, using default size 1024
WARN [KafkaSource] Avg event size for partition tianjin-smart-city-1:1 not available, using default size 1024
WARN [KafkaSource] Avg event size for partition tianjin-smart-city-1:2 not available, using default size 1024
WARN [KafkaSource] Avg event size for partition tianjin-smart-city-1:3 not available, using default size 1024
ERROR [MRJobLauncher] Collected 0 task states while expecting 4 task states
Exception in thread "main" java.lang.IllegalArgumentException: Missing required property writer.staging.dir
        at com.google.common.base.Preconditions.checkArgument(Preconditions.java:93)
        at gobblin.util.WriterUtils.getWriterStagingDir(WriterUtils.java:55)
        at gobblin.util.JobLauncherUtils.cleanStagingData(JobLauncherUtils.java:172)
        at gobblin.runtime.AbstractJobLauncher.cleanupStagingData(AbstractJobLauncher.java:460)
        at gobblin.runtime.AbstractJobLauncher.launchJob(AbstractJobLauncher.java:285)
        at gobblin.runtime.mapreduce.CliMRJobLauncher.run(CliMRJobLauncher.java:60)
        at org.apache.hadoop.util.ToolRunner.run(ToolRunner.java:70)
        at gobblin.runtime.mapreduce.CliMRJobLauncher.main(CliMRJobLauncher.java:133)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:606)
        at org.apache.hadoop.util.RunJar.run(RunJar.java:221)
        at org.apache.hadoop.util.RunJar.main(RunJar.java:136)
```
And my job configuration file is listed:
```
job.name=Pull-From-Kafka-To-HDFS
job.group=Kafka
job.description=Kafka Extractor for Gobblin
job.lock.enabled=false

writer.staging.dir=/tianjin/task-staging

source.class=gobblin.source.extractor.extract.kafka.KafkaSimpleSource
converter.classes=gobblin.converter.IdentityConverter
extract.namespace=gobblin.extract.kafka

topic.whitelist=tianjin-smart-city-1
topic.name=tianjin-smart-city-1
kafka.brokers=kafka-1:9092

writer.builder.class=gobblin.writer.SimpleDataWriterBuilder
bootstrap.with.offset=earliest

writer.partition.level=hourly
writer.partition.pattern=YYYY/MM/dd/HH

mr.job.max.mappers=500
```
I have checked DEBUG level log, but find no information helps to address the problem.

**Comment by zliu41** [comment]
Hi @Evangileon I wasn't able to reproduce your problem. Can you let us know your sysconfig file, as well as what properties are in the `State` object when it throws that exception?

**Comment by liyinan926** [comment]
@Evangileon are you using the script `gobblin-mapreduce.sh` to run the job? Can you share with us the command you used and the environment variables you set?

**Comment by zhaolmtravelsky** [comment]
I encounter the same probleam. My command is bin/gobblin-mapreduce.sh --jt ip:port --fs hdfs://xxx --workdir hdfs://xxx/tmp/test/gobblin_work --conf /home/data/test/gobbin_config/kafka_hdfs.pull。

job.name=PullFromKafka
job.group=Kafka
job.description=Kafka Extractor for Gobblin
job.lock.enabled=false

source.class=gobblin.source.extractor.extract.kafka.KafkaSimpleSource

#extract.namespace=gobblin.extract.kafka

writer.destination.type=HDFS
#writer.output.format=AVRO
writer.fs.uri=hdfs://xxx

data.publisher.type=gobblin.publisher.TimePartitionedDataPublisher

topic.whitelist=test
bootstrap.with.offset=earliest

kafka.brokers=ip1:port1,ip2:port2


writer.partition.level=hourly
writer.partition.pattern=YYYY/MM/dd/HH
writer.builder.class=gobblin.writer.SimpleDataWriterBuilder
#writer.builder.class=gobblin.writer.AvroTimePartitionedWriterBuilder
writer.file.path.type=tablename
writer.partition.column.name=header.time

mr.job.max.mappers=20

extract.limit.enabled=true
extract.limit.type=time
#(other possible values: rate, count, pool)
extract.limit.time.limit=15
extract.limit.time.limit.timeunit=minutes


Can anyone help me?

**Comment by zliu41** [comment]
@zhaolmtravelsky you don't have `writer.staging.dir` in your kafka_hdfs.pull. Is it in another config file?

**Comment by Evangileon** [comment]
@zliu41 @liyinan926 My command was `gobblin-mapreduce.sh --conf pipeline.pull`.
I have tried define `writer.staging.dir` explicitly in job configuration file, but it never worked.
After I modified `yarn-site.xml` and `mapred-site.xml` files to minimum configuration, I addressed the issue. Because configuration of Java heap size is too large, so that worker instance failed at initialization. Therefore Node Manager could not get the configuration object serialized from master node. It just reported some variable missing.
